pubtools/pulplib/_impl/model/repository/file.py

```diff
@@ -20,3 +23,53 @@ class FileRepository(Repository):
     )
 
     mutable_urls = attr.ib(default=attr.Factory(lambda: ["PULP_MANIFEST"]), type=list)
+
+    def upload_file(self, filename, relative_url=None):
```

Comment:
It just occurred to me now that this API forces you into only being to upload content if you've already got it in a file on disk.

I'd expect that this may not even meet your current use-case all that well.  You're creating a JSON document, right?  Probably using Python "json" module?

So, with this API, don't you have to manipulate a dict in memory, use json to write it out to some file, then use this API to read it back from the same file again?  The data is written to disk and immediately read back?

Would it make more sense to let this take an IO object as well?  What I mean is:

```
def upload_file(self, file_obj, relative_url=None):
```

- if file_obj is a string, it's the name of a file to open.
- otherwise, if file_obj is a "file object" https://docs.python.org/3/library/io.html, data is read from that object

Could prevent you having to write content to disk just to ask this library to read it again.

Reply:
yeah, makes sense. I'll make the change.